internal/cluster/job.go

```diff
@@ -0,0 +1,109 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	batchv1 "k8s.io/api/batch/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// CleanJob cleans up a job in a given namespace
+func CleanJob(client crclient.Client, name string, namespace string) error {
+	job, err := getJobInNamespace(client, name, namespace)
+	if err != nil {
+		return err
+	}
+	err = deleteJob(client, job)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getJobInNamespace finds a job with a given name in a namespace
+func getJobInNamespace(client crclient.Client, name string, namespace string) (*batchv1.Job, error) {
+	job := &batchv1.Job{}
+	err := client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, job)
+	if err != nil {
+		return job, err
+	}
+	return job, nil
+}
+
+// deleteJob deletes a given job and cleans up any pods associated with it
+func deleteJob(c crclient.Client, job *batchv1.Job) error {
+	deleteForeground := metav1.DeletePropagationBackground
+	err := c.Delete(context.TODO(), job, &crclient.DeleteOptions{PropagationPolicy: &deleteForeground})
+	if err != nil {
+		log.Error(err, "Error deleting job: "+job.Name)
+		return err
+	}
+	//we don't use job.Selector here since it's autogenerated and contains only generated controller-uid unique for each job instance
+	//if we use job-name which is also generated label but not used in selector - we clean up pods from old jobs
+	pods, err := GetPodsBySelector(c, job.Namespace, "job-name="+job.Name)
+	if err != nil {
+		return err
+	}
+	for _, pod := range pods.Items {
+		log.Info(fmt.Sprintf("Job '%s/%s' related pod %s is not removed automatically", job.Namespace, job.Name, pod.Name))
+		err = DeletePod(c, job.Namespace, pod.Name)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Wait for the job to complete. Times out if the job isn't complete after $(timeout) seconds
+func WaitForJobCompletion(client crclient.Client, name string, namespace string, timeout time.Duration) error {
+	const interval = 1 * time.Second
+	return wait.PollImmediate(interval, timeout, func() (bool, error) {
+		job, err := getJobInNamespace(client, name, namespace)
+		if err != nil {
+			return false, err
+		}
+
+		if job.Status.Succeeded > 0 {
```

Comment:
Probably a future improvement, but I wonder if it's worth it to check `job.Status.Failed > 0` as well and log a line mentioning it.